client-spark/spark-2/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java

```diff
@@ -142,7 +142,12 @@ public boolean unregisterShuffle(int appShuffleId) {
     }
     // For Spark driver side trigger unregister shuffle.
     if (lifecycleManager != null) {
-      lifecycleManager.unregisterAppShuffle(appShuffleId);
+      if (celebornConf.clientFetchThrowsFetchFailure()) {
```

Comment:
WDYT,  if we use `def unregisterAppShuffle(appShuffleId: Int, hasSubShuffleIds: Boolean):` to support both cases 

```suggestion
        def unregisterAppShuffle(appShuffleId: Int, hasSubShuffleIds: Boolean): Unit = {
    logInfo(s"Unregister appShuffleId $appShuffleId starts...")
    appShuffleDeterminateMap.remove(appShuffleId)
    if (hasSubShuffleIds) {
      val shuffleIds = shuffleIdMapping.remove(appShuffleId)
      if (shuffleIds != null) {
        shuffleIds.synchronized(
          shuffleIds.values.map {
            case (shuffleId, _) =>
              unregisterShuffle(shuffleId)
          })
      }
    } else {
      unregisterShuffle(appShuffleId)
    }
  }
      
```

Reply:
per my perspective, I'm not fond of adding a new definition of SubShuffleIds since it was not used anywhere else
Regarding appShuffleDeterminateMap cleanup, I prefer to put it in SparkShuffleManager. unregisterShuffle(), because it is added in SparkShuffleManager.registerShuffle()

Reply:
@FMX @waitinfuture WDYT?

Reply:
IMO, I think a coherent method "unregisterAppShuffle" would be better. But I think there is no need to introduce a new concept called "hasSubShuffleIds", maybe "hasMapping" could be enough to tell others that the shuffle ID has mapping.

Reply:
@RexXiong, `unregisterAppShuffle` method makes sense to me, because the implementation of this method is doing the unregistration of `appShuffleId`. Meanwhile, what's the definition of `subShuffleId`. There is no concept for `subShuffleId` but the mapping between `appShuffleId` and `celebornShuffleId`.

Reply:
```suggestion
def unregisterAppShuffle(appShuffleId: Int, hasShuffleMapping: Boolean)
```

Reply:
How about call `unregisterShuffle` if `shuffleIdMapping` doesn't contain `appShuffleId`?
```
  def unregisterAppShuffle(appShuffleId: Int): Unit = {
    logInfo(s"Unregister appShuffleId $appShuffleId starts...")
    val shuffleIds = shuffleIdMapping.remove(appShuffleId)
    if (shuffleIds != null) {
      shuffleIds.synchronized(
        shuffleIds.values.map {
          case (shuffleId, _) =>
            unregisterShuffle(shuffleId)
        })
    } else {
      unregisterShuffle(appShuffleId)
    }
  }
```

cc @RexXiong @ErikFang @FMX @SteNicholas

Reply:
After further discussion with @RexXiong @FMX , I think @FMX 's suggestion is better, `unregisterShuffle` can be made private then. WDYT @ErikFang

Reply:
I'm ok with the idea
will have LifecycleManager.(appShuffleId: Int, hasMapping: Boolean) to wrap the logic internally with hasMapping as a flag. it will be very similar to @RexXiong 's suggestion